File: 0860-design-circular-queue/0860-design-circular-queue.java
class MyCircularQueue {
    int[] queue;
    int front = 0, rear = -1, len = 0;

    public MyCircularQueue(int k) {
        this.queue = new int[k];
    }
    
    public boolean enQueue(int value) {
        if (!this.isFull()) {
            this.rear = (this.rear + 1) % this.queue.length;
            this.queue[rear] = value;
            this.len++;
            
            return true;
        } else return false;
    }
    
    public boolean deQueue() {
        if (!this.isEmpty()) {
            this.front = (this.front + 1) % this.queue.length;
            this.len--;

            return true;
        } else return false;
    }
    
    public int Front() {
        return this.isEmpty() ? -1 : this.queue[this.front];
    }
    
    public int Rear() {
        return this.isEmpty() ? -1 : this.queue[this.rear];
    }
    
    public boolean isEmpty() {
        return this.len == 0;
    }
    
    public boolean isFull() {
        return this.len == this.queue.length;
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue obj = new MyCircularQueue(k);
 * boolean param_1 = obj.enQueue(value);
 * boolean param_2 = obj.deQueue();
 * int param_3 = obj.Front();
 * int param_4 = obj.Rear();
 * boolean param_5 = obj.isEmpty();
 * boolean param_6 = obj.isFull();
 */

File: 0860-design-circular-queue/0860-design-circular-queue.kt
class MyCircularQueue(k: Int) {
    var queue: IntArray = IntArray(k)
    var front = 0
    var rear = -1
    var len = 0

    fun enQueue(value: Int): Boolean {
        return if (!isFull()) {
            this.rear = (this.rear + 1) % this.queue.size
            this.queue[rear] = value
            this.len++
            true
        } else false
    }

    fun deQueue(): Boolean {
        return if (!isEmpty()) {
            this.front = (this.front + 1) % this.queue.size
            this.len--
            true
        } else false
    }

    fun Front(): Int {
        return if (this.isEmpty()) -1
        else queue[front]
    }

    fun Rear(): Int {
        return if (this.isEmpty()) -1
        else queue[rear]
    }

    fun isEmpty(): Boolean {
        return this.len == 0
    }

    fun isFull(): Boolean {
        return this.len == this.queue.size
    }

}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * var obj = MyCircularQueue(k)
 * var param_1 = obj.enQueue(value)
 * var param_2 = obj.deQueue()
 * var param_3 = obj.Front()
 * var param_4 = obj.Rear()
 * var param_5 = obj.isEmpty()
 * var param_6 = obj.isFull()
 */